src/compute/src/compute_state.rs

```diff
@@ -660,8 +662,45 @@ impl PendingPeek {
             cursor.step_key(&storage);
         }
 
+        let peek = &mut self.peek;
+        let oks = self.trace_bundle.oks_mut();
+        match oks {
+            SpecializedTraceHandle::Bits32Row(key_types, oks_handle) => {
+                Self::collect_finished_data(
+                    peek,
+                    oks_handle,
+                    key_types,
+                    &Vec::default(),
+                    max_result_size,
+                )
+            }
+            SpecializedTraceHandle::RowRow(oks_handle) => Self::collect_finished_data(
+                peek,
+                oks_handle,
+                &Vec::default(),
+                &Vec::default(),
+                max_result_size,
+            ),
+        }
+    }
+
+    /// Collects data for a known-complete peek.
+    fn collect_finished_data<K, V>(
+        peek: &mut Peek<Timestamp>,
+        oks_handle: &mut TraceRowHandle<K, V, Timestamp, Diff>,
+        key_types: &Vec<ColumnType>,
+        val_types: &Vec<ColumnType>,
+        max_result_size: u32,
+    ) -> Result<Vec<(Row, NonZeroUsize)>, String>
+    where
+        K: Columnation + Data + FromRowByTypes + for<'a> IntoRowByTypes<'a>,
+        V: Columnation + Data + for<'a> IntoRowByTypes<'a>,
```

Comment:
Is it correct that `val_types` is always `Vec::new()` and `V` is always `Row`? And we also don't plan to introduce specialized value types in the scope of this work? If so I would simplify things by hardcoding values to `Row`. We can always add the genericity later when we need it.

Reply:
I'd like to try to add support for empty values, if possible. The analysis in the [initial results internal doc](https://www.notion.so/materialize/Prototyping-Arrangement-Specialization-2866676a2d7d453fbaa802fa64ea8864) indicates this could be pretty helpful. In that case, the schema for the values would still be empty, but we'd hopefully get it from type inference. And we'd need to have the generic parameters for values anyway. So we might as well leave things ready for more value specializations?

Reply:
Yep, makes sense!